factor/sms/classes/factor.php

```diff
@@ -0,0 +1,358 @@
+                   AND revoked = 0';
+
+        return $DB->get_records_sql($sql, [$user->id, $this->name]);
+    }
+
+    /**
+     * SMS Factor implementation.
+     *
+     * {@inheritDoc}
+     */
+    public function is_enabled() {
+        global $CFG;
+        // If local_aws is not installed, not enabled.
+        if (!file_exists($CFG->dirroot . '/local/aws/version.php')) {
+            return false;
+        } else {
+            return parent::is_enabled();
+        }
+    }
+
+    /**
+     * SMS Factor implementation.
+     *
+     * {@inheritDoc}
+     */
+    public function has_input() {
+        return true;
+    }
+
+    /**
+     * SMS Factor implementation.
+     *
+     * {@inheritDoc}
+     */
+    public function has_setup() {
+        return true;
+    }
+
+    /**
+     * SMS Factor implementation
+     *
+     * {@inheritDoc}
+     */
+    public function show_setup_buttons() {
+        global $DB, $USER;
+        // If there is already a factor setup, don't allow multiple (for now).
+        $sql = 'SELECT *
+                  FROM {tool_mfa}
+                 WHERE userid = ?
+                   AND factor = ?
+                   AND secret = ?
+                   AND revoked = 0';
+
+        $record = $DB->get_record_sql($sql, [$USER->id, $this->name, '']);
+        return !empty($record) ? false : true;
+    }
+
+    /**
+     * SMS Factor implementation.
+     *
+     * {@inheritDoc}
+     */
+    public function has_revoke() {
+        return true;
+    }
+
+    /**
+     * Generates and sms' the code for login to the user, stores codes in DB.
+     *
+     * @return int the instance ID being used.
+     */
+    private function generate_and_sms_code() {
+        global $DB, $USER;
+
+        $duration = get_config('factor_sms', 'duration');
+        $secret = $this->secretmanager->create_secret($duration, false);
+        $instance = $DB->get_record('tool_mfa', ['factor' => $this->name, 'userid' => $USER->id, 'revoked' => 0]);
+
+        // There is a new code that needs to be sent.
+        if (!empty($secret)) {
+            // Grab the singleton SMS record.
+            $this->sms_verification_code($secret, $instance->label);
+        }
+        return $instance->id;
+    }
+
+    /**
+     * This function sends an SMS code to the user based on the instanceid provided.
+     * If rawcode is provided, it will send the raw code through to the phone number in the user profile.
+     *
+     * @param int $secret the secret to send.
+     * @param int|null $rawcode If provided, a raw code to send to the users profile phone no.
+     * @return void
+     */
+    private function sms_verification_code($secret, $phonenumber) {
+        global $CFG, $SITE;
+
+        // Here we should get the information, then construct the message.
+        $content = ['site' => $SITE->fullname, 'url' => $CFG->wwwroot, 'code' => $secret];
+        $message = get_string('smsstring', 'factor_sms', $content);
```

Comment:
I think the url can be simplified to just the domain without the http and also without any moodle path suffix. The spec is a little vague on this, but I think less cruft in the message is better